gapis/api/vulkan/primeable_image_data.go

```diff
@@ -564,163 +565,219 @@ func (p *imagePrimer) newPrimeableImageData(img VkImage, opaqueBoundRanges []VkI
-			storeToStagingImgKits, err := kb.BuildStoreKits(p.sb, storeToStagingImgRecipes...)
-			if err != nil {
-				return nil, log.Errf(p.sb.ctx, err, "failed to build store kits from recipes")
-			}
-			staging := &ipPrimeableStoreKits{
-				img:   stagingImg.VulkanHandle(),
-				queue: queue.VulkanHandle(),
-				kits:  storeToStagingImgKits,
-			}
-			// TODO: layout transition for old state image pre image store to staging image
-			err = staging.prime(p.sb,
-				useSpecifiedLayout(VkImageLayout_VK_IMAGE_LAYOUT_UNDEFINED),
-				useSpecifiedLayout(ipStoreImageLayout),
-			)
-			if err != nil {
-				return nil, log.Errf(p.sb.ctx, err, "failed at storing data to staging image")
-			}
-			// TODO: layout transition for old state image post iamge store to staging image
-			kits, err := kb.BuildStoreKits(p.sb, recipes...)
-			if err != nil {
-				return nil, log.Errf(p.sb.ctx, err, "failed at build store kits for priming storage image")
-			}
-			primeable.kits = kits
-			return primeable, nil
 		}
+		kits, err := kb.BuildStoreKits(p.sb, recipes...)
+		if err != nil {
+			return nil, log.Errf(p.sb.ctx, err, "failed to build store kits from recipes")
+		}
+		primeable.kits = kits
+		return primeable, nil
 	}
 
 	primeByPreinitialization := (!primeByCopy) && (!primeByRendering) && (!primeByImageStore) && (oldStateImgObj.Info().Tiling() == VkImageTiling_VK_IMAGE_TILING_LINEAR) && (oldStateImgObj.Info().InitialLayout() == VkImageLayout_VK_IMAGE_LAYOUT_PREINITIALIZED)
 	if primeByPreinitialization {
-		if fromHostData {
-			queue := getQueueForPriming(p.sb, oldStateImgObj, VkQueueFlagBits_VK_QUEUE_TRANSFER_BIT|VkQueueFlagBits_VK_QUEUE_GRAPHICS_BIT|VkQueueFlagBits_VK_QUEUE_COMPUTE_BIT)
-			if queue.IsNil() {
-				return nil, log.Errf(p.sb.ctx, nilQueueErr, "[Building primeable image data that can be primed by preinitialization with host data, image: %v]", img)
-			}
-			return &ipPrimeableByPreinitialization{p: p, img: img, opaqueBoundRanges: opaqueBoundRanges, queue: queue.VulkanHandle()}, nil
-		} else {
-			return nil, log.Errf(p.sb.ctx, notImplErr, "[Building primeable image data that can be primed by preinitialization with device data, image: %v]", img)
+		queue := getQueueForPriming(p.sb, oldStateImgObj, VkQueueFlagBits_VK_QUEUE_TRANSFER_BIT|VkQueueFlagBits_VK_QUEUE_GRAPHICS_BIT|VkQueueFlagBits_VK_QUEUE_COMPUTE_BIT)
+		if queue.IsNil() {
+			return nil, log.Errf(p.sb.ctx, nilQueueErr, "[Building primeable image data that can be primed by preinitialization with host data, image: %v]", img)
 		}
+		return &ipPrimeableByPreinitialization{p: p, img: img, opaqueBoundRanges: opaqueBoundRanges, queue: queue.VulkanHandle()}, nil
+
 	}
 	return nil, log.Errf(p.sb.ctx, nil, "No way build primeable image data for image: %v", img)
 }
+
+// newPrimeableImageDataFromDevice builds primeable image data from the source image, which is on the device already.
+func (p *imagePrimer) newPrimeableImageDataFromDevice(srcImg, dstImg VkImage) (primeableImageData, error) {
```

Comment:
We will need to handle sparse images here as well.
I suggest, for now, we simply check and fail if we come across one.